sklearn/preprocessing/tests/test_encoders.py

```diff
@@ -637,3 +638,207 @@ def test_categories(density, drop):
 @pytest.mark.parametrize('Encoder', [OneHotEncoder, OrdinalEncoder])
 def test_encoders_has_categorical_tags(Encoder):
     assert 'categorical' in Encoder()._get_tags()['X_types']
+
+
+@pytest.mark.parametrize("is_sparse", [True, False])
+@pytest.mark.parametrize("drop", [None, "first"])
+def test_one_hot_encoder_pd_categories(is_sparse, drop):
+    pd = pytest.importorskip('pandas')
+
+    X_df = pd.DataFrame({
+        'col_str': ['a', 'b', 'b', 'a'],
+        'col_int': [3, 2, 1, 2]}, columns=['col_str', 'col_int'])
+
+    str_category = pd.api.types.CategoricalDtype(
+         categories=['b', 'a'], ordered=True)
```

Comment:
You could also add a category not present in the data

Reply:
BTW, I think doing it like this can be cleaner:

```
X_df = pd.DataFrame({
    'col_str': pd.Categorical(['a', b', 'b', 'a'], categories=['a', 'b'], ordered=True),
    ....
    )
```

instead of first creating the dataframe, then the dtypes and then astyping.